### ISSUE TITLE
1.2.0 : (side effect?) MQTT v3 async client connect listener token::on_success() is called twice when using set_callback and connect with cb.

### ISSUE THREAD

**Issue by Pascal-Fremaux** [eclipse/paho.mqtt.cpp#322]
After integration of the latest Paho C / CPP, we have encountered an issue: the (token) connected on_success callback is called twice.
So we have had to patch to ignore the second event.

After investigation, it seems it came from a recent change in Paho CPP 1.2.0:
https://github.com/eclipse/paho.mqtt.cpp/commit/c1f48fd7deccda1324c3534b79c162229cb7227c

As you can see in the log below (MQTT Paho C logs are the [MQTT Min]), the others are from our client)
Paho C emits first a completion for the connect::on_success callback, then it completes the (mqtt) connected callback
The first calls our callback ConnectListener::on_success() (as expected), via the token::on_success in Paho CPP (calling listener->on_success(*this); ). 
This is the expected behavior that was working in Paho CPP 1.1.0.

The second completes the Paho CPP callback async_client::on_connected() that was supposed to call the client connected callback if any 
cb->connected(cause_str);  
, which is empty in our code, but now in Paho CPP also additionally calls a second time the token connect on_success() (so our ConnectListener::on_success() callback is called a second time) 
tok->on_success(nullptr);

PLNX: 2021-01-25 09:28:34.480 D 13877 13931 TRANSPOR [MQTT Min]: 20210125 092834.480 (2600457984)  (1)< MQTTAsync_completeConnection:1933 (0)
PLNX: 2021-01-25 09:28:34.480 D 13877 13931 TRANSPOR [MQTT Min]: 20210125 092834.480 Connect succeeded to ssl://localhost:8883
PLNX: 2021-01-25 09:28:34.480 D 13877 13931 TRANSPOR [MQTT Min]: 20210125 092834.480 Calling connect success for client octavdevivc1
PLNX: 2021-01-25 09:28:34.480 D 13877 13930 TRANSPOR [MQTT Min]: 20210125 092834.480 (2608850688)  (1)< Thread_wait_cond:436 (0)
PLNX: 2021-01-25 09:28:34.480 D 13877 13931 TRANSPOR ConnectListener::on_success()
PLNX: 2021-01-25 09:28:34.480 D 13877 13930 TRANSPOR [MQTT Min]: 20210125 092834.480 (2608850688)  (1)> MQTTAsync_checkTimeouts:1637
PLNX: 2021-01-25 09:28:34.480 D 13877 13931 TRANSPOR [MQTT Min]: 20210125 092834.480 (2600457984)  (1)> MQTTProperties_copy:433
PLNX: 2021-01-25 09:28:34.480 D 13877 13931 TRANSPOR [MQTT Min]: 20210125 092834.480 (2600457984)  (1)< MQTTProperties_copy:442
PLNX: 2021-01-25 09:28:34.480 I 13877 13931 TRANSPOR ConnectListener::operator()(sessionPresent: 1)
PLNX: 2021-01-25 09:28:34.480 V 13877 13931      STM Sending message 3
PLNX: 2021-01-25 09:28:34.480 V 13877 13931 Transpor submitting task 3 
PLNX: 2021-01-25 09:28:34.480 D 13877 13931 TRANSPOR [MQTT Min]: 20210125 092834.480 (2600457984)  (1)> MQTTProperties_free:401
PLNX: 2021-01-25 09:28:34.480 D 13877 13931 TRANSPOR [MQTT Min]: 20210125 092834.480 (2600457984)  (1)< MQTTProperties_free:424
PLNX: 2021-01-25 09:28:34.480 V 13877 13890 Transpor executing task "" [3/3] with latency (ms) = 0
PLNX: 2021-01-25 09:28:34.480 V 13877 13890      STM handleMessage: msg.what=3
PLNX: 2021-01-25 09:28:34.480 D 13877 13890      STM [Mqtt] {connecting}: <onConnected>
PLNX: 2021-01-25 09:28:34.480 D 13877 13890 TRANSPOR No need for MQTT subscribe -> FSM connected
PLNX: 2021-01-25 09:28:34.480 I 13877 13890 TRANSPOR MqttHelper::notifyConnected()
PLNX: 2021-01-25 09:28:34.480 I 13877 13890 XCPR_MOC XCP Time Event : OCM / OCM_MQTT_CONN_ESTABLISHED
PLNX: 2021-01-25 09:28:34.480 D 13877 13931 TRANSPOR [MQTT Min]: 20210125 092834.480 Calling connected for client octavdevivc1
PLNX: 2021-01-25 09:28:34.480 D 13877 13890     CORE [SM FSM] ->onSessionStateChange(connected)
PLNX: 2021-01-25 09:28:34.480 V 13877 13890      STM Sending message 0
PLNX: 2021-01-25 09:28:34.480 D 13877 13931 TRANSPOR ConnectListener::on_success()

As far as I understand, the duplication come from a side effect as we use async_client::set_callback(cb) in addition of the async_client::connect(connect_options opts, void* userContext, iaction_listener& cb)

The set_callback, in addition to call Paho C MQTTAsync_setCallbacks(), also call MQTTAsync_setConnected (not new) but now with the modified async_client::on_connected, so the side effect.

**Comment by Pascal-Fremaux** [comment]
In addition I tried to convert our usage of the async_client::set_callback(), and use instead the 2 APIs :
void set_connection_lost_handler(connection_handler cb)  // similar on_connection_lost
void set_message_callback(message_handler cb)   //  similar on_message_arrived

and it seems to work well !

so I discourage the usage of the old set_callback() in the async_client when using the connect with cb...

**Comment by fpagliughi** [comment]
Yes, a problem with the library is that it has been evolving over the years, and all the legacy API's still exist. I didn't mean to change the behavior, so I will have a look at that aspect of it, but there are several ways to be notified of events, so perhaps one thing that needs to be done is better documentation.

**Comment by fpagliughi** [comment]
OK. The behavior changed to fix a bug, #300, in which the `reconnect()` call wasn't being signaled as complete. That definitely needs to happen, but it should _not_ call `on_success()` twice. I bet there's a lot of code that relies on it running only once per token.

In general, I agree that `set_callback()` is old and there are better ways to use the API, but I can't remove anything until a major version upgrade, and until then I need to make sure everything keeps working as expected.

For now I'll mark this as a bug and take a closer look when I have a few minutes (hours, days...) to work on it.

**Comment by NANDLAB** [comment]
I have the same bug.
My callback class looks like this:
```cpp
class mqtt_node_paho : public mqtt::callback, public mqtt::iaction_listener
```
An instance of this class is given to both `set_callback()` and to the `connect()` method.
When connected, the `on_success()` callback is called **twice** and then `connected()` is called. For me this is not a critical problem, because you can just ignore `on_success()` and only use `connected()`, but the current behavior is unexpected.

**Comment by avivruham** [comment]
I encounter the same issue with v5.
Do you have an estimation for the next version with a fix?

Thanks

**Comment by fpagliughi** [comment]
This is fixed with Paho C v1.3.13
